### PR TITLE
CVE-2007-4559 Patch

### PR DESCRIPTION
**Pull request by TrellixVulnTeam**
# Patching CVE-2007-4559

Hi, we are security researchers from the Advanced Research Center at [Trellix](https://www.trellix.com). We have began a campaign to patch a widespread bug named CVE-2007-4559. CVE-2007-4559 is a 15 year old bug in the Python tarfile package. By using extract() or extractall() on a tarfile object without sanitizing input, a maliciously crafted .tar file could perform a directory path traversal attack. We found at least one unsantized extractall() in your codebase and are providing a patch for you via pull request. The patch essentially checks to see if all tarfile members will be extracted safely and throws an exception otherwise. We encourage you to use this patch or your own solution to secure against CVE-2007-4559. Further technical information about the vulnerability can be found in this [blog](https://www.trellix.com/en-us/about/newsroom/stories/research/tarfile-exploiting-the-world.html).

If you have further questions you may contact us through this projects lead researcher [Kasimir Schulz](mailto:kasimir.schulz@trellix.com).

### CODE REVIEW

**Review by hawflau — DISMISSED** [review]
Thanks @TrellixVulnTeam for the security patch!

I believe we should look to move `is_within_directory` and `safe_extract` to a  common place to reduce repetition. Also should we add some tests?

**Review by hnnasit — APPROVED** [review]
LGTM!

File: aws_lambda_builders/utils.py [hunk 1]
```diff
@@ -196,3 +196,28 @@ def create_symlink_or_copy(source: str, destination: str) -> None:
             exc_info=ex if LOG.isEnabledFor(logging.DEBUG) else None,
         )
         copytree(source, destination)
+
+
+def _is_within_directory(directory, target):
```
Comment (lucashuy) [inline aws_lambda_builders/utils.py:201]: nit: can we add type hinting for the params and return value for these new methods?

File: tests/functional/test_utils.py [hunk 2]
```diff
@@ -63,6 +63,22 @@ def test_must_return_valid_go_architecture(self):
         self.assertEqual(get_goarch(""), "amd64")
 
 
+class TestExtractTarFile(TestCase):
+    def test_extract_tarfile_unpacks_a_tar(self):
```
Comment (lucashuy) [inline tests/functional/test_utils.py:67]: question: is testing the exception (ie actually performing the attack) something that should be done here?

Comment (hawflau) [reply tests/functional/test_utils.py:67]: good catch. I intended but forgot to add the path traversal case. Will add it.

File: aws_lambda_builders/utils.py (hunk 1 above)
Comment (hawflau) [reply aws_lambda_builders/utils.py:201]: sure, can do

**Review by lucashuy — APPROVED** [review]

**Pull request merged**